cirq/circuits/expectation_value.py

```diff
@@ -0,0 +1,340 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Expectation value tool for pauli operators given a circuit"""
+
+# imports
+from typing import Dict, Union
+
+import numpy as np
+from cirq.google import XmonSimulator, XmonMeasurementGate
+from cirq.circuits import Circuit
+from cirq.ops import RotXGate, RotYGate, MeasurementGate, Pauli, PauliString
+from cirq.devices import GridQubit
+from cirq.line import LineQubit
+from cirq.ops import NamedQubit
+
+
+# Helper Function to add measurement gates to qubits
+def measurement_gates(circuit: Circuit,
+                      q_dict: Dict[Union[NamedQubit, GridQubit, LineQubit],
+                                   Union[int, str]]):
+    """
+    circuit: cirq.Circuit to add measurements to
+    q_dict: dictionary of qubits, keys are integers or strings,
+        values are corresponding qubits
+
+    return: circuit with measurements added
+    """
+    meas = [XmonMeasurementGate(key=index).on(
+        qubit) for qubit, index in q_dict.items()]
+
+    circuit.append(meas)
+
+    return circuit
+
+
+def expectation_from_sampling(circuit: Circuit,
+                              operator: Dict[PauliString, float],
+                              n_samples: int=100,
+                              quadratic_z: bool =False):
+    """
+    Calculates the expactation value of cost function (operator)
+
+    Args:
+        circuit: circ.Circuit type which prepares the state we are interested
+                 in calculating the expectation value of operator
+
+        operator: Operator input as a dictionary where keys are
+                    cirq.ops.PauliString type and values are coefficients
+                  example:
+                  paulistring = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.X, qubits[1]: (cirq.Pauli.Z)})
+                  paulistring2 = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.Z, qubits[2]: (cirq.Pauli.Y)})
+                  operator = {paulistring: 1.234, paulistring2: -5.678}
+
+                sampling is for future incorporation with QPU (not simulator)
+
+        n_samples: number of runs for sampling
+
+        quadratic_z: Boolean that determines whether operator only has terms
```

Comment:
Can this be computed cheaply directly from the operator? If so, that would be a way to avoid having an option that the user can get wrong.

Otherwise I would rename this to "assume_quadratic".